velox/functions/sparksql/Arithmetic.h

```diff
@@ -25,6 +25,43 @@
 
 namespace facebook::velox::functions::sparksql {
 
+template <typename T>
+struct RoundFunction {
+  template <typename TNum, typename TDecimals, bool alwaysRoundNegDec = false>
+  FOLLY_ALWAYS_INLINE TNum
+  round(const TNum& number, const TDecimals& decimals = 0) {
+    static_assert(
+        !std::is_same_v<TNum, bool> && "round not supported for bool");
+
+    if constexpr (std::is_integral_v<TNum>) {
+      if constexpr (alwaysRoundNegDec) {
+        if (decimals >= 0)
+          return number;
+      } else {
+        return number;
+      }
+    }
+    if (!std::isfinite(number)) {
+      return number;
+    }
+
+    double factor = std::pow(10, decimals);
+    static const TNum kInf = std::numeric_limits<TNum>::infinity();
+    if (number < 0) {
+      return (std::round(std::nextafter(number, -kInf) * factor * -1) /
+              factor) *
+          -1;
+    }
+    return std::round(std::nextafter(number, kInf) * factor) / factor;
```

Comment:
This probably won't work with something like `round(0.49999999999999994, 0)`.  Let's do it correctly by
1. `sprinf` to a `char[]`
2. Move the decimal point to correct position
3. `sscanf` back

Reply:
Thanks, I will try this way.

Reply:
Hi @Yuhta , I create a example code, but result is not expected. 0.575 will be convert to 0.57499999999999996 If we use sprintf. 
```
#include <iostream>
#include <cmath>
#include <limits>
using namespace std;
int main(){
 char buffer[40];
 double num = 0.49999999999999994;
 sprintf(buffer,"%.17f",num);
 string sNum = buffer;
 std::cout << "sNum : " << sNum << std::endl;

 double num1 = 0.575;
 char buffer1[40];
 sprintf(buffer1,"%.17f",num1);
 string sNum1 = buffer1;
 std::cout << "sNum1 : " << sNum1 << std::endl;
 return 0;
}

```
![image](https://github.com/facebookincubator/velox/assets/14065538/f55a4778-8af7-465d-8005-1b3d6d247971)

Actually, I think cases like 0.575 are quite common. and This situation `0.49999999999999994`
is very rare.

Reply:
You need to check how Spark implement it and do the same then.  Both `round(0.49999999999999994, 0)` and `round(0.575, 2)` has to be correct and one is not rarer than the other.